File: java/121.best-time-to-buy-and-sell-stock.java
/*
 * @lc app=leetcode id=121 lang=java
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int profit = 0;
        for(int i = 1; i < prices.length; i++) {
            profit += prices[i] - prices[i-1];
            maxProfit = Math.max(maxProfit, profit);
            if(profit < 0) profit = 0;
        }

        return maxProfit;
    }
}
// @lc code=end

File: cpp/121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int max = 0;
        int profit = 0;

        auto prev = prices.begin();

        for(auto curr = std::next(prev); curr != prices.end(); ++curr) {
            profit += *curr - *prev;
            profit = std::max(profit, 0);
            max = std::max(max, profit);
            prev = curr;
        }

        return max;
    }
};
// @lc code=end

